libs/grout/README.md

```diff
@@ -0,0 +1,160 @@
+    }
+    await store.updatePersonAge(input.name, input.age);
+    return ok();
+  },
+});
+
+const app = express();
+
+// When registering the API with an Express app, you must use a separate
+// Router. That way you can isolate the Grout API routes from other routes in
+// your app.
+const router = express.Router();
+
+// You must have the router use the `express.text()` middleware with type
+// `application/json` exactly as shown below.
+router.use(express.text({ type: 'application/json' }));
+
+// Then, simply register the API routes.
+grout.registerRoutes(router, api);
+
+// Finally, mount the router on your app. You can use any path prefix you want -
+// here we use `/api`. It's important that you don't use any other body-parsing
+// middleware upstream of the router - e.g. don't call `app.use(express.json())`
+// before this.
+app.use('/api', router);
+
+// Don't forget to export the API type - we'll need it for the client.
+export type MyApi = typeof api;
+```
+
+### Client
+
+```ts
+import * as grout from '@votingworks/grout';
+
+// First, we need to import the API type from the server. This will *not* import
+// any of the actual server code - just the type definition.
+import type { MyApi } from './server';
+
+// We need to set the base URL for the client. Make sure to include the path
+// prefix (in our case, /api).
+const baseUrl = 'http://localhost:1234/api';
+
+// Create the client using MyApi as a type parameter.
+const apiClient = grout.createClient<MyApi>({ baseUrl });
+
+// Now we can call the API methods we defined as normal functions.
+await apiClient.getAllPeople(); // => [{ name: 'Alice', age: 99 }, ...]
+await apiClient.getPersonByName({ name: 'Bob' }); // => { name: 'Bob', age: 42 }
+await apiClient.updatePersonAge({ name: 'Bob', age: 43 }); // => ok()
+
+// If we make a mistake, we'll get a type error, just like with normal typed functions
+await apiClient.getAllPeeple(); // => TS error: Property getAllPeeple does not exist
+await apiClient.getPersonByName({ nam: 'Bob' }); // => TS error: Argument of type '{ nam: string; }' is not assignable to parameter of type '{ name: string; }'.
+await apiClient.updatePersonAge({ name: 'Bob', age: '1' }); // => TS error: Argument of type '{ name: 'Bob'; age: '1'; }' is not assignable to parameter of type '{ name: string; age: number; }'.
+
+// Since we used the Result type, known error cases must be handled explicitly.
+const updateResult = await apiClient.updatePersonAge({ name: 'Bob', age: -1 });
+if (updateResult.isErr()) {
+  console.error(updateResult.error.message); // => 'Age must be at least 1.'
+}
+
+// If there's an unexpected server error (e.g. a crash or runtime exception),
+// the Promise returned by the client method will be rejected, so we still need
+// to handle exceptions.
+try {
+  await apiClient.getAllPeople();
+} catch (error) {
+  if (error instanceof grout.ServerError) {
+    console.error(error.message);
+  }
+}
```

Comment:
I take it that, in practice, we'll be relying on a global error handler like the react-query error handler for these errors instead of wrapping every `apiClient` call in a try-catch - is that thinking correct?

Reply:
Yes, I'm planning to look into setting up a React error boundary for unexpected errors.